File: src/DiagToolsValidator/DiagToolsValidator.Core/Functionality/DotNetFunctionality/DotNetTool.fs
namespace DiagToolsValidator.Core.Functionality

open System.IO
open System.Net.Http
open System.Collections.Generic


module DotNetTool =
    let GetToolIL (toolRoot: string) (toolName: string) (toolVersion: string) =
        let intermediateDirectoryRoot = Path.Combine(toolRoot, ".store", toolName, toolVersion, toolName, toolVersion, "tools")
        try
            let intermediateDirectory = 
                Directory.GetDirectories(intermediateDirectoryRoot, "net*")
                |> Array.head
            Choice1Of2 (Path.Combine(intermediateDirectory, "any", $"{toolName}.dll"))
        with ex -> 
            ex.Data.Add("DotNetTool.GetToolDll", $"Fail to find IL for {toolName} in {toolRoot}")
            Choice2Of2 ex


    let InstallDotNetTool (dotNetEnv: Dictionary<string, string>)
                          (toolRoot: string)
                          (toolFeed: string)
                          (configFilePath: string) 
                          (toolVersion: string) 
                          (toolName: string) =
        let argument = 
            if Path.Exists(configFilePath)
            then $"tool install {toolName} --tool-path {toolRoot} --version {toolVersion} --add-source {toolFeed} --configfile {configFilePath}"
            else $"tool install {toolName} --tool-path {toolRoot} --version {toolVersion} --add-source {toolFeed}"
        DotNet.RunDotNetCommand dotNetEnv 
                                argument
                                ""
                                true
                                CommandLineTool.PrinteOutputData
                                CommandLineTool.PrintErrorData
                                true


    let DownloadPerfcollect (perfcollectPath: string) =
        let perfcollectUrl = "https://raw.githubusercontent.com/microsoft/perfview/main/src/perfcollect/perfcollect"
        task {
            use httpClient = new HttpClient()
            let! response = httpClient.GetAsync(perfcollectUrl)
            response.EnsureSuccessStatusCode() |> ignore
            let! streamReader = response.Content.ReadAsStreamAsync()
            let streamWriter = File.OpenWrite(perfcollectPath)
            
            streamReader.CopyToAsync(streamWriter) |> ignore
            return perfcollectPath
        }
        |> Async.AwaitTask
        |> Async.Catch
        |> Async.RunSynchronously
        

File: src/DiagToolsValidationRunner/DiagToolsValidationRunner.Core/Functionality/DotNetTool.cs
namespace DiagToolsValidationRunner.Core.Functionality
{
    public static class DotNetTool
    {
        public static string GetToolIL(string toolRoot, string toolName, string toolVersion)
        {
            string frameworkRoot = Path.Combine(
                toolRoot, ".store", toolName, toolVersion, toolName, toolVersion, "tools");
            string frameworkDir =
                Directory.GetDirectories(frameworkRoot, "net*")
                    .First();
            string toolIL = Path.Combine(frameworkDir, "any", $"{toolName}.dll");

            if (!File.Exists(toolIL))
            {
                throw new Exception($"{nameof(DotNetTool)}: Tool IL {toolIL} doesn't exist");
            }

            return toolIL;
        }

        public static CommandInvokeResult InstallDotNetTool(Dictionary<string, string> dotNetEnv,
                                                       string toolRoot,
                                                       string toolFeed,
                                                       string toolVersion,
                                                       string toolName,
                                                       string? configFilePath=null,
                                                       bool redirectStdOutErr = true,
                                                       bool silent = false)
        {
            string dotNetExecutable = DotNetInfrastructure.GetDotNetExecutableFromEnv(dotNetEnv);
            string argument =
                configFilePath switch
                {
                    null => $"tool install {toolName} --tool-path {toolRoot} --version {toolVersion} --add-source {toolFeed}",
                    _ => $"tool install {toolName} --tool-path {toolRoot} --version {toolVersion} --add-source {toolFeed} --configfile {configFilePath}"
                };

            CommandInvoker invoker = new(dotNetExecutable, argument, dotNetEnv, "", redirectStdOutErr, silent);
            return invoker.WaitForResult();
        }
    }
}
